### PR TITLE
Handle implicit object creation in inline temp. refactoring

### LINKED ISSUE #50207: Inline temporary variable produce invalid code with the new instantiation shorthand.
Hello, I don't know if this is the correct repo because I don't know if this codefix is a Roslyn Analyzer or a Visual Studio thing.

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.101](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

image

The inlined code should be "new Closure(args).Run" and not "new(args).Run".

### Steps To Reproduce

```csharp
class MyClass
{
    void Test()
    {
        MyClass myClass = new();
        myClass.ToString();
    }
}
```

### Expected behavior

```csharp
class MyClass
{
    void Test()
    {
        new MyClass().ToString();
    }
}
```

### Actual behavior
```csharp
class MyClass
{
    void Test()
    {
        MyClass myClass = new();
            (new()).ToString();
    }
}
```

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #50207

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
If you do fix this, please make sure other untyped exprs work as well (like inlining 'null/default/some-lambda/etc.').

**Comment by Youssef1313** [comment]
@CyrusNajmabadi These are already tested:

https://github.com/dotnet/roslyn/blob/51548ef1a906f84faa0f39fc6cdb177c5a33893a/src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs#L2300-L2322

https://github.com/dotnet/roslyn/blob/51548ef1a906f84faa0f39fc6cdb177c5a33893a/src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs#L1612-L1638

I'm waiting for the CI to fail to push a fix.

File: src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs [hunk 1]
```diff
@@ -5426,5 +5426,34 @@ public void Bar()
...
+            var code = @"
+class MyClass
+{
+    void Test()
+    {
+        MyClass [||]myClass = new();
+        myClass.ToString();
+    }
+}
+
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/CodeActions/InlineTemporary/InlineTemporaryTests.cs:5443]:
```suggestion
```

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs [hunk 2]
```diff
@@ -461,6 +461,13 @@ private static ExpressionSyntax SkipRedundantExteriorParentheses(ExpressionSynta
             var localSymbol = (ILocalSymbol)semanticModel.GetDeclaredSymbol(variableDeclarator, cancellationToken);
             var newExpression = InitializerRewriter.Visit(expression, localSymbol, semanticModel);
 
+            if (newExpression.IsKind(SyntaxKind.ImplicitObjectCreationExpression))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:464]: I think it's worth a comment to explain what's going on

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
thanks!

**Pull request merged**